### ISSUE TITLE
What are the best practices for get UPN?

### ISSUE THREAD

**Issue by pa-chihaya-katagi** [microsoftconnect/ms-intune-app-sdk-android#258]
## Questions

I am currently using the following code to get the UPN.
```
MAMComponents.get(MAMUserInfo.class).getPrimaryUser()
```

It is not annotated with a deprecated annotation, but the documentation seems to indicate that this is not the recommended method.
https://microsoftconnect.github.io/ms-intune-app-sdk-android/reference/com/microsoft/intune/mam/policy/MAMUserInfo.html

Checking the alternative method, it is an API to get the OID.
I have confirmed that I can get the OID, but how do I get the UPN?

Thank you in advance.

## Details

- **Intune Android App SDK Version:**
  - 11.0.0
- **Android Device Make and Model:**
- **Android Device OS Version:**
- **Android Studio Version:**
  - Android Studio Ladybug | 2024.2.1 Patch 3

**Comment by manupa-ms** [comment]
MAM SDK API methods that take UPNs to specify identities are deprecated since MAM SDK 10.2.1. They will be removed completely at the next major version increment as UPN is not a reliable parameter to uniquely identify a user account - it can change overtime; two different accounts could share the same UPN and so on. In order to avoid inconsistencies relating to UPNs, this change has been introduced. 
New methods that specify identities by OID (also known as AAD User ID, AAD ID or Entra ID) should be used instead. Most of the SDK's identity APIs use the provided OID as the identifier for the identity. The MAM SDK APIs will return the OID string as the identity and will require the OID string parameter for the identity. Some methods may also take or return a UPN string, in which case the UPN is for informational purposes only.
Please refer: https://github.com/microsoftconnect/ms-intune-app-sdk-android/discussions/238

**Comment by pa-chihaya-katagi** [comment]
@manupa-ms 
Thank you for your response.

My app is determining if the email address entered by the user in the app matches the UPN that is retrieved from the MAM SDK.
So I need the UPN, not the OID.

> Some methods may also take or return a UPN string, in which case the UPN is for informational purposes only.

I would appreciate a link to the JavaDoc for this method.
Also, will this method be removed in a future version?

**Comment by manupa-ms** [comment]
The docs reference is here: https://microsoftconnect.github.io/ms-intune-app-sdk-android/reference/com/microsoft/intune/mam/client/identity/[MAMPolicyManager](https://microsoftconnect.github.io/ms-intune-app-sdk-android/reference/com/microsoft/intune/mam/client/identity/MAMPolicyManager.html).html

All methods returning UPN strings would be deprecated in a future version.

**Comment by pa-chihaya-katagi** [comment]
@manupa-ms 
Thank you.
So all APIs involved in UPN are deprecated.

Is there a way to map OIDs to UPNs in Android using the MSAL API, as in the iOS answer here?
https://github.com/microsoftconnect/ms-intune-app-sdk-ios/issues/470#issuecomment-2372514969

**Comment by microsoftjoe** [comment]
Hi @pa-chihaya-katagi, we don't recommend using UPNs for account identifiers, since they are not suitable unique IDs. An account can have multiple UPNs, and the MAM SDK won't generally know all of them. It is also possible in the Entra Identity system for different accounts to have the same UPN. For those reasons, the MAM SDK has switched to using the OID of the account as the unique identifier internally, so we can ensure that the correct MAM policies are used for a given account.
In the case that the MAM SDK does know about multiple UPNs for a given identity, the above-mentioned API method will return one of them, but we don't guarantee which one.
If you wanted to retrieve an account from the OID and get the username property to find the UPN, you could call getAccount():
https://javadoc.io/static/com.microsoft.identity.client/msal/2.2.3/com/microsoft/identity/client/IMultipleAccountPublicClientApplication.html#getAccount-java.lang.String-
I can't give MSAL or Entra support in general, but the docs are all public, so I recommend reading through them and you might find more info about what you need.

**Comment by pa-chihaya-katagi** [comment]
@microsoftjoe 
Thank you very much. I will try the API you mentioned.

**Comment by pa-chihaya-katagi** [comment]
@microsoftjoe @manupa-ms 
Sorry after it was closed.

I tried the API you gave me, I can get it if I am already logged in with M365, but not when I am not logged in.
Does this API require me to log in to M365 before I can get the UPN?
```
        PublicClientApplication.createMultipleAccountPublicClientApplication(
                ApplicationManager.self(),
                R.raw.auth_config_single_account_debug,
                new IPublicClientApplication.IMultipleAccountApplicationCreatedListener() {

                    @Override
                    public void onCreated(IMultipleAccountPublicClientApplication application) {
                        Thread workerThread = new Thread(() -> {
                            try {
                                List<IAccount> list = application.getAccounts();
                                Log.d("test",
                                      "list.get(0).getUsername(): " + list.get(0).getUsername());
                                Log.d("test",
                                      "list.get(0).getId(): " + list.get(0).getId());
```
I found a way to get the UPN using the OID before logging in with the Microsoft Graph API, but is this not possible with the MSAL API?

I am able to get UPN even before login with this method on iOS.
https://github.com/microsoftconnect/ms-intune-app-sdk-ios/issues/470#issuecomment-2598575450

**Comment by microsoftjoe** [comment]
Hi @pa-chihaya-katagi, MAM doesn't have any interest in accounts that have not logged into an app and been registered with the MAM SDK. I really can't advise you any further than I have above, since the inner workings of MSAL are not part of my area of expertise. I don't think this is a MAM question anymore.

**Comment by pa-chihaya-katagi** [comment]
@microsoftjoe 
Yes, I apologize for the question being beyond the scope of MAM. Thank you very much.
